### ISSUE TITLE
Incorrect or no URL when entering new data

### ISSUE THREAD

**Issue by bondjimbond** [digitalutsc/authority-link-auto-complete#1]
If you do the following:

* Enter some data into the Alternate Link Text field that matches the data on the lookup target, which produces a URI in the URI field, and then
* Erase that text and enter something new that does not match data on the target

The old URI persists and saves with the new (invalid) text.

Also: if you enter some Alternate Link Text that does not match the authority, and no URI is produced, when you save your content the entire field is discarded with no error.

1. Is there a way to discard the URI when typing new and invalid data into the Alternate Link Text lookup field?
2. Is there a way to validate the data in the Alternate Link Text lookup field?
3. Can the form return an error when there is text in the Alternate Link Text field but nothing in the URI field?

**Comment by kstapelfeldt** [comment]
What if we want to point to a URI but have a different label? Sometimes we slightly adjust terms for capitalization or number to match the rest of the repository vocabulary, but still want to link to the authority to indicate that this is the "thing" we mean in an internet context. Let me know if I'm missing something in what you're reporting (totally possible)

**Comment by bondjimbond** [comment]
In the instance where this came up for me and I was testing it...

- I started typing "data", which brought up "dataset" in my target vocabulary. This was selected, and a URI was generated.
- I realized that wasn't the term I wanted to use, erased it, and put in an entirely different term (not in the vocabulary).
- Because the new term was not in the target vocabulary, the "dataset" URI remained in the URI field and was saved.

I guess there is no perfect way to handle these things.

**Comment by kstapelfeldt** [comment]
Yeah that does seem like an error, but I can't think about how validation would work in a way that allows me to use a 'non recommended' term or that works in a generalized way. Maybe we could stick up a warning for users: "The text in your label has changed. Please verify that your URI remains accurate" prior to saving...would that help? (and is it possible @ravence ?) I'm curious - What happens if you clear out the URI and label, and start typing your label again? @ravence curious about your thoughts too.

**Comment by bondjimbond** [comment]
> What happens if you clear out the URI and label, and start typing your label again?

In that case it behaves as normal. Note though, as mentioned in my first post, that if you type out a term that is not in the target vocabulary and no URI is generated, no data is saved in that field.

So I guess if your target vocab's term is "dataset" but you wanted to use "Datasets" with the same URI, you might not actually retrieve that URI (and therefore not save the term at all) unless you start with "dataset" and then adjust.

**Comment by bondjimbond** [comment]
How common is the use case of using an authority but changing it locally? It's one I hadn't thought about; in my own use case, I want all my sites using exactly the same terms with no deviations.

**Comment by larkgz** [comment]
> Maybe we could stick up a warning for users: "The text in your label has changed. Please verify that your URI remains accurate" prior to saving

The Javascript is already watching to see when the autocomplete is selected (that’s how it fills in the URI field), so checking if the label and URI match whenever the label changes should be doable. I like the idea of a tooltip or text notification of some kind rather than disabling the functionality outright since we seem to be using it that way. But I'm not sure how common either of these use cases are (I'm just a developer) so I will refer back to the two of you.

> Also: if you enter some Alternate Link Text that does not match the authority, and no URI is produced, when you save your content the entire field is discarded with no error.

It seems like this issue pertains to the Controlled Access Terms module. If there's no URI and the field is optional and you enter Alternate Link Text, then the field is discarded as described.

**Comment by kstapelfeldt** [comment]
"Just" is not a word I would apply to you in any context @ravence ! FWIW, I like the idea of a tooltip. @bondjimbond re: use cases, I don't know how common this is, but it's the use case at UTSC so that's what we're developing against. We want the freedom to use the URI even if I'm not using the orthographically identical 'preferred term.' @bondjimbond would warning of a mismatch resolve your issue if we can wrangle it?

**Comment by bondjimbond** [comment]
> would warning of a mismatch resolve your issue if we can wrangle it?

It's a good patch. There's still the question of what to do with terms that are entered with no corresponding term at the endpoint. Do you want to discard these because they don't match your vocabulary, or do you want to accept them and write new terms to the vocabulary (a REST API function perhaps)? 

At minimum: 

> If there's no URI and the field is optional and you enter Alternate Link Text, then the field is discarded as described.

Is it possible to add some functionality where, if the user attempts to save the form when there is content in the Alternate Link Text field but no content in the URI field, it interrupts you with a warning to fix your data (rather than saving but discarding what the user entered)?

**Comment by larkgz** [comment]
> Is it possible to add some functionality where, if the user attempts to save the form when there is content in the Alternate Link Text field but no content in the URI field, it interrupts you with a warning to fix your data (rather than saving but discarding what the user entered)?

I can try adding a hook to do that here, but I think it would make more sense to fix this particular issue in the controlled access terms module since it occurs even when this module is not installed. Unless there is a particular reason that this functionality should be added specifically here that I'm missing?

**Comment by bondjimbond** [comment]
> Unless there is a particular reason that this functionality should be added specifically here that I'm missing?

No specific reason except that I ran into the problem when trying out this module. If the problem is coming from a different module, it'd best be fixed there.
